dll/ntdll/nt_0600/ldr/ldrnotify.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * PROJECT:     ReactOS NT Layer/System API
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     DLL Load Notification Implementation
+ * COPYRIGHT:   Copyright 2024 Ratin Gao <ratin@knsoft.org>
+ */
+
+#include "ntdll_vista.h"
+
+typedef struct _LDR_DLL_NOTIFICATION_ENTRY
+{
+    LIST_ENTRY List;
+    PLDR_DLL_NOTIFICATION_FUNCTION Callback;
+    PVOID Context;
+} LDR_DLL_NOTIFICATION_ENTRY, *PLDR_DLL_NOTIFICATION_ENTRY;
+
+static RTL_STATIC_LIST_HEAD(LdrpDllNotificationList);
+
+/* Initialize critical section statically */
+
+static RTL_CRITICAL_SECTION LdrpDllNotificationLock;
+
+static RTL_CRITICAL_SECTION_DEBUG LdrpDllNotificationLockDebug = {
+    .CriticalSection = &LdrpDllNotificationLock
+};
+
+static RTL_CRITICAL_SECTION LdrpDllNotificationLock = {
+    &LdrpDllNotificationLockDebug,
+    -1,
+    0,
+    0,
+    0,
+    0
+};
+
+NTSTATUS
+NTAPI
+LdrRegisterDllNotification(
+    _In_ ULONG Flags,
+    _In_ PLDR_DLL_NOTIFICATION_FUNCTION NotificationFunction,
+    _In_opt_ PVOID Context,
+    _Out_ PVOID *Cookie)
+{
+    PLDR_DLL_NOTIFICATION_ENTRY NewEntry;
+
+    /* Check input parameters */
+    if (Flags != 0 || NotificationFunction == NULL || Cookie == NULL)
+    {
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    /* Allocate new entry and assign input values */
+    NewEntry = RtlAllocateHeap(LdrpHeap, 0, sizeof(*NewEntry));
+    if (NewEntry == NULL)
+    {
+        return STATUS_NO_MEMORY;
+    }
+    NewEntry->Callback = NotificationFunction;
+    NewEntry->Context = Context;
+
+    /* Add node to the end of global list */
+    RtlEnterCriticalSection(&LdrpDllNotificationLock);
+    if (LdrpDllNotificationList.Blink->Flink != &LdrpDllNotificationList)
+    {
+        FatalListEntryError(LdrpDllNotificationList.Blink->Blink,
```

Comment:
Hmmm..... Not sure why we would need to explicitly have these checks in the code.
It's as if, the very same linked-list insert/remove... macros as those present in the WDK (see XDK rtlfuncs.h) headers were used also in user-mode, with the `NO_KERNEL_LIST_ENTRY_CHECKS` macro disabled, in the ntdll code...